### PR TITLE
Using user's account language to set campaign params

### LINKED ISSUE #5091: Global English campaign signup transactional email
When a user signs up for a campaign that is translated in Global English, s/he should receive a campaign signup transactional email in Global English. This is needed because content could potentially be different in |STEP_ONE| and |STEP_THREE|

Also note that we need to make sure we're pulling in the proper fields for this since technically the Global English content is in different fields than the US English fields.

related to #4956 

**Code of interest**:
https://github.com/DoSomething/phoenix/blob/ffb2ee968fc45dc5854ccb03c3dbf993eb38fe05/lib/modules/dosomething/dosomething_signup/dosomething_signup.module#L585-L601

**Called from**:
- **dosomething_signup_get_mbp_params**: https://github.com/DoSomething/phoenix/blob/ffb2ee968fc45dc5854ccb03c3dbf993eb38fe05/lib/modules/dosomething/dosomething_signup/dosomething_signup.module#L578

then:
- **dosomething_signup_mbp_request**: https://github.com/DoSomething/phoenix/blob/ffb2ee968fc45dc5854ccb03c3dbf993eb38fe05/lib/modules/dosomething/dosomething_signup/dosomething_signup.module#L534

then:
- **dosomething_signup_get_mbp_params**: https://github.com/DoSomething/phoenix/blob/ffb2ee968fc45dc5854ccb03c3dbf993eb38fe05/lib/modules/dosomething/dosomething_signup/dosomething_signup.module#L534-L537

### PR DESCRIPTION
**Pull request by blisteringherb**
- When user's sign up for a campaign, they will now receive their signup
  emails with copy in their language of choice.
- The user's language should always be set appropriately either when they create an account or when they change their own language settings.

Fixes #5091 and #4956

### CODE REVIEW

**Comment by angaither** [comment]
code looks :ok: 
but I don't think we are allowing users to change their own lang? what if this isn't set?

**Comment by blisteringherb** [comment]
The language by default when a user creates an account will be 'en'. We'll need a hook_user_insert to make sure the language gets set properly based on the Fastly headers, so the user's language should always be set properly. If the hook_user_insert malfunctions for some reason then the user will get content in 'en'. We may want to log when the user's language setting is different than what we get from Fastly.

**Comment by itsjoekent** [comment]
Yeah this was the point I raised yesterday @angaither 

https://dosomething.slack.com/archives/global/p1442416380000776

We decided it's safe to assume when a user registers they've always have the correct language set

**Comment by angaither** [comment]
...right but what about users we have already w/o a lang?

**Comment by itsjoekent** [comment]
what users dont have a language set? dont they have to have at least english set?

**Comment by blisteringherb** [comment]
@angaither @deadlybutter Hm, it looks like in the db the language column is empty or und for all but about 6 users. Should we run a script now before we turn on global to set all of those users' language to 'en'?  Is there any advantage to keeping the language empty or 'und' for a user?

**Comment by itsjoekent** [comment]
I don't see any advantage, I think we should put everyone on the same system. @angaither ?

**Comment by angaither** [comment]
@blisteringherb rather than writing a script to handle this, I have sort of done it here
https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_global/dosomething_global.module#L116

**Comment by blisteringherb** [comment]
@angaither That function accounts for users with global admin accounts but not others. We could expand that function or write some other helpers to return the language based on logic, but I think it would be more performant if we make sure the user's language is set properly in hook_user_insert and then generally use the user's language whenever we need to show them something.

**Comment by angaither** [comment]
@blisteringherb that sounds good to me, all I meant was we should have a default of english if the user doesn't have a language set on insert (which I think is the case for most of our users now)

**Comment by itsjoekent** [comment]
according to morgans comment above, only 6 users don't have a language set. can we set it for them & assume from thereon out all users will have a language set?

**Comment by angaither** [comment]
@deadlybutter, he said all but 6 have a lang set. 
I just ran the query ~20 k have something set
the other 3 million have nothing or `und` set

**Comment by itsjoekent** [comment]
OH. well that changes things. 

EDIT: so yeah, we should probably make a helper function which we call to grab a users language that can provide a default fallback. (should this be global-en, or en?)

**Comment by mikefantini** [comment]
since the majority of our current users are :us: and the ratio of new users will most likely stay majority :us: for a while I would say default should be en.

**Comment by mshmsh5000** [comment]
:+1:. Because of #5130, the system's default language may become Global English. Makes sense for users still to default to US English, though.

**Comment by deezone** [comment]
@blisteringherb @mikefantini There might need to be clarification on this issue as my understanding is a campaign sign up should trigger a transactional email in the **language of the campaign** NOT in the language of the user (as the issue description is suggesting).

Looking at the code changes in this PR it looks like you are using the language of the campaign. Just wanted to clarify as the bulk of the comments on this issue are discussing the user's language preference at the time of signup.

**Comment by mikefantini** [comment]
@blisteringherb @DeeZone the languages of the emails should be as follows:

| Type of Email | Language |
| --- | --- |
| User Registration | Language of user |
| PW Reset | Language of user |
| Campaign signup | Language of campaign |
| Campaign reportback | Language of campaign |

Let me know if any further clarification is needed.

**Comment by blisteringherb** [comment]
@DeeZone @mikefantini I'm unsure what you mean by "language of campaign". The user's language preference or the language from the URL are the only languages that we can use. A campaign will have multiple languages.

**Comment by mikefantini** [comment]
@blisteringherb take the following case:
- Brazilian user
- signs up for node 123
- node 123 has a US English translation published and a Global English translation published but has no BR-PT translation published
- in this case even though the user's language setting would be PT-BR, because she signed up for a Global English translation of the campaign, her transactional email should be in Global English, not BR-PT
- user then signs up for node 124
- node 124 has a US English translation published, a Global English translation published, and a BR-PT translation published
- in this case, the user would be seeing the BR-PT translation of the campaign and would then get a transactional email for node 124 in BR-PT

does that make sense/seem doable @blisteringherb?

File: lib/modules/dosomething/dosomething_global/dosomething_global.module [hunk 1]
```diff
@@ -221,4 +221,37 @@ function dosomething_global_convert_country_to_language($country) {
...
+ * Get the appropriate language for the application.
+ *
+ * @param $account
+ *   The user's account
+ * @param  $node
+ *   Optional node.
+ *
+ */
+function dosomething_global_get_language($account, stdClass $node = NULL) {
+  if ($node = $options['node']) {
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module]: `$options['node']` isn't getting passed in?

File: lib/modules/dosomething/dosomething_global/dosomething_global.module [hunk 2]
```diff
@@ -221,4 +221,37 @@ function dosomething_global_convert_country_to_language($country) {
...
+    elseif ($node->translations->data['en-global']['status']) {
+      $language = 'en-global';
+    }
+    // Default to node language if there's no translation in their language and no 'en-global'.
+    else {
+      $language = $node->language;
+    }
+  }
+  elseif ($language = $account->language) {
+    return $language;
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:251]: shouldn't this say so you aren't returning in the middle?

``` php
elseif (isset($account->language)) {
 $language = $account->language
}
```

Comment (blisteringherb) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:251]: Is there a reason to avoid returning here?

Comment (mshmsh5000) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:251]: This is an assignment, not a comparison. `if ($language = $account->language)` will always be `true`.

If the intended logic is "if the current language is equal to the account language, then just return the language here and skip the last few lines," that makes sense to me.

Comment (angaither) [inline lib/modules/dosomething/dosomething_global/dosomething_global.module:251]: :+1: ya agree, I think I just misunderstood that

**Pull request merged**